### ISSUE TITLE
[labs/ssr] altering state/property of SSR'd element before firstUpdate breaks reactivity

### ISSUE THREAD

**Issue by hrmcdonald** [lit/lit#3971]
### Which package(s) are affected?

SSR (@lit-labs/ssr)

### Description

When an element is SSR'd if any property is changed before the first time `firstUpdated` is called the reactivity of that property is broken from that point onward. This behavior occurs when a `@state` or `@property` value is changed in `connectedCallback` which I could maybe understand and work around since I could see that getting called before hydration occurs...but it also happens if the value is changed during `willUpdate` which is a Lit lifecycle hook so that doesn't seem to be the case. 

> I cannot verify this, but I'm pretty sure these scenarios used to work with an older version of Lit SSR too so it might be a regression?

This is just kind of a big issue because sometimes we need to execute checks that we can't do on the server during `willUpdate`. Things like evaluating slotted content. It just kind of breaks the whole component if SSR'd making it unusable. 

I don't feel like waiting until a second render is a really great solution here either. For one there's not a very ergonomic way to do that out of the box at the moment, and it means there's a good second or two of the component flashing the wrong state before the component is hydrated and renders a _second_ time.

### Reproduction

There is kind of a wide range of scenarios that can cause this that you can in this example. Even other variables not changed before `firstUpdated` won't be evaluated properly within expressions that happen to also reference one that was.

Recreation in Astro on stackblitz:
[https://stackblitz.com/edit/github-gpznzw-uwksmx](https://stackblitz.com/edit/github-gpznzw-uwksmx?file=src%2Fcomponents%2Fmy-element.ts)

```
export class MyElement extends LitElement {
  static styles = css`
    .text-red {
      color: red;
    }
  `;

  @property({ type: Boolean, reflect: true }) red = false;

  @state() makeRed = false;

  @state() trueAfterFirstUpdate = false;

  connectedCallback() {
    super.connectedCallback();

    // Comment this out and `red` works as intended after firstUpdated
    this.red = true;
  }

  willUpdate(changes) {
    super.willUpdate(changes);

    // Comment this out and `makeRed` works as intended after firstUpdated
    if (!isServer) {
      this.makeRed = true;
    }
  }

  firstUpdated() {
    // When these are called before `firstUpdated` setting them to true or even false and then later true has no effect.
    // They just seem completely disconnected from the reactive update cycle when changed before this point.
    this.red = false;
    this.makeRed = false;
    // requestUpdates or performUpdates don't make a difference
    this.requestUpdate('red', false);
    this.requestUpdate('makeRed', false);

    setTimeout(() => {
      this.red = true;
      this.makeRed = true;
      this.requestUpdate('red', false);
      this.requestUpdate('makeRed', false);
    }, 1000);

    this.trueAfterFirstUpdate = true;
  }

  render() {
    // Whether these print true or false here, the template will not take changes into account if changed before `firstUpdated`.
    console.log(this.red, this.makeRed);

    // Using a classMap or just string interpolation makes no difference.
    // Even when a variable that was not changed before firstUpdated is used, that expression stops getting evaluated properly when one that was is included
    return html`
      <div class="${
        this.red || this.makeRed ? 'text-red' : null
      }">should be red</div>
      <div class=${classMap({
        'text-red': this.red || this.makeRed,
      })}>should also be red</div>
      <div class=${classMap({
        'text-red': this.trueAfterFirstUpdate || this.makeRed,
      })}>somehow this also isn't red</div>
      <div class="${
        this.red ? 'text-red' : null
      }">This will update after the 2nd/3rd render beacuse after firstUpdated it is toggled to false and then again later to true.</div>
      <div class="${
        this.makeRed ? 'text-red' : null
      }">When changed during willUpdate on the first render though that is not the case...</div>
    `;
  }
}

customElements.define('my-element', MyElement);
```

### Workaround

There is no workaround other than creating some solution to not have any logic anywhere that does anything with reactive state/properties until after `firstUpdated` is called first, even within other Lit lifecycle hooks. However, as we expressed, this really degrades the user experience and almost makes SSR not worth it in a lot of cases, but then because of how we'd need to write components it'd degrade non-SSR'd experiences too.

### Is this a regression?

No or unsure. This never worked, or I haven't tried before.

### Affected versions

Failing in `3.1.x`, If it worked it was on or before `2.3.0`

### Browser/OS/Node environment

Browser: any
OS: MacOS
Node: 18
NPM: 9.5.1

**Comment by hrmcdonald** [comment]
Not 100% sure if this is related, but here is an example that might showcase this same bug expressed through even the boolean attribute setter `?` prefix:

[https://stackblitz.com/edit/github-gpznzw?file=src%2Fcomponents%2Fwrapped-element.ts](https://stackblitz.com/edit/github-gpznzw?file=src%2Fcomponents%2Fwrapped-element.ts)

In this example when a boolean attribute is set via a camelCase prop it behaves just like one of the properties in my original example above. The underlying property gets set to `true`, but the first render is as if it is `false`/doesn't exist. However attempts to force it to re-evaluate with something like `requestUpdate('camelCase`, false)` don't do anything.

If the prop name in the template where the prop is being set is made lowercase though, it works? But non-lowercased it works fine when not-SSR'd.

If this is a different issue though, we can open up a separate issue ticket.

**Comment by augustjk** [comment]
The boolean attribute does look like a separate problem. Filed https://github.com/lit/lit/issues/3972

**Comment by augustjk** [comment]
After digging into the original issue at hand, it's not that those properties are necessarily losing their reactivity, it's just that when hydration happens, it trusts the existing values as the "current" value and stores then on the parts without updating the DOM. The value here refers to the value of what the "class" attribute should be for each div, i.e. "text-red" in the example.

Subsequent update requests with the same value doesn't do anything because Lit thinks the value haven't changed. The reactivity can be checked by selecting the element in the inspector and setting the prop directly with `$0.red = false` etc. The key is to make sure the actual value of the classname changes as the cache is based on that instead of individual prop/states.

Ultimately, a strategy like described here https://github.com/lit/lit/issues/1434#issuecomment-1601410576 would alleviate this by making sure to actually update the DOM with the current value, instead of just trusting that it's the same from server render time.

**Comment by hrmcdonald** [comment]
Yeah we figured it was ultimately due to some kind of caching going on that was preventing further changes from updating things, but weren't sure what the root cause was specifically within the hydration process. This should hopefully help us somehow come up with a workaround for now until the direction for that other thread is finalized. thanks!
